ofono/src/gprs.c

```diff
@@ -893,6 +895,140 @@ static DBusMessage *pri_get_properties(DBusConnection *conn,
+			return TRUE;
+		}
+	}
+	return FALSE;
+}
+
+static void pri_reset_context_properties(DBusConnection *conn,
+	struct pri_context *ctx, const struct ofono_gprs_provision_data *ap)
+{
+	struct ofono_gprs *gprs = ctx->gprs;
+	gboolean changed = FALSE;
+
+	if (strcmp(ctx->context.apn, ap->apn)) {
+		changed = TRUE;
+		strcpy(ctx->context.apn, ap->apn);
+		ofono_dbus_signal_property_changed(conn, ctx->path,
+			OFONO_CONNECTION_CONTEXT_INTERFACE, "AccessPointName",
+					DBUS_TYPE_STRING, &ap->apn);
+	}
+
+	if (ap->name && strncmp(ctx->name, ap->name, MAX_CONTEXT_NAME_LENGTH)) {
+		changed = TRUE;
+		strncpy(ctx->name, ap->name, MAX_CONTEXT_NAME_LENGTH);
+		ofono_dbus_signal_property_changed(conn, ctx->path,
+			OFONO_CONNECTION_CONTEXT_INTERFACE, "Name",
+					DBUS_TYPE_STRING, &ap->name);
+	}
+
+	if (pri_update_str(ctx->context.username, ap->username)) {
+		changed = TRUE;
+		ofono_dbus_signal_property_changed(conn, ctx->path,
+			OFONO_CONNECTION_CONTEXT_INTERFACE, "Username",
+					DBUS_TYPE_STRING, &ap->username);
+	}
+
+	if (pri_update_str(ctx->context.password, ap->password)) {
+		changed = TRUE;
+		ofono_dbus_signal_property_changed(conn, ctx->path,
+			OFONO_CONNECTION_CONTEXT_INTERFACE, "Password",
+					DBUS_TYPE_STRING, &ap->password);
+	}
+
+	if (ap->type == OFONO_GPRS_CONTEXT_TYPE_MMS) {
+		if (pri_update_str(ctx->message_proxy, ap->message_proxy)) {
+			changed = TRUE;
+			ofono_dbus_signal_property_changed(conn, ctx->path,
+				OFONO_CONNECTION_CONTEXT_INTERFACE,
+					"MessageProxy", DBUS_TYPE_STRING,
+							&ap->message_proxy);
+		}
+
+		if (pri_update_str(ctx->message_center, ap->message_center)) {
+			changed = TRUE;
+			ofono_dbus_signal_property_changed(conn, ctx->path,
+				OFONO_CONNECTION_CONTEXT_INTERFACE,
+					"MessageCenter", DBUS_TYPE_STRING,
+							&ap->message_center);
+		}
+	}
+
+	if (gprs->settings && changed) {
+		write_context_settings(gprs, ctx);
+		storage_sync(gprs->imsi, SETTINGS_STORE, gprs->settings);
+	}
+}
+
+static gboolean pri_valid_ap(const struct ofono_gprs_provision_data *ap)
+{
+	if (!ap->apn || strlen(ap->apn) > OFONO_GPRS_MAX_APN_LENGTH ||
+							!is_valid_apn(ap->apn))
+		return FALSE;
+
+	if (ap->username &&
+			strlen(ap->username) > OFONO_GPRS_MAX_USERNAME_LENGTH)
+		return FALSE;
+
+	if (ap->password &&
+			strlen(ap->password) > OFONO_GPRS_MAX_PASSWORD_LENGTH)
+		return FALSE;
+
+	if (ap->message_proxy &&
+			strlen(ap->message_proxy) > MAX_MESSAGE_PROXY_LENGTH)
+		return FALSE;
+
+	if (ap->message_center &&
+			strlen(ap->message_center) > MAX_MESSAGE_CENTER_LENGTH)
+		return FALSE;
+
+	return TRUE;
+}
+
+static DBusMessage *pri_reset_properties(DBusConnection *conn,
+					DBusMessage *msg, void *data)
+{
+	struct pri_context *context = data;
+	struct ofono_gprs *gprs = context->gprs;
+	struct ofono_modem *modem = __ofono_atom_get_modem(gprs->atom);
+	struct ofono_sim *sim = __ofono_atom_find(OFONO_ATOM_TYPE_SIM, modem);
+	struct ofono_gprs_provision_data *settings;
+	int i, count;
+
```

Comment:
I could implement re-provisioning entirely on the client side - just parse mbpi file, deactivate the context and apply the properties. In fact, I have already done that and I didn't like it at all, it's too complex and fragile (too many steps involved + race condition at every step). Doing it on the ofono side is much easier, more reliable and allows to avoid duplicating the logic of selecting settings from mbpi (like, provider name matching).

Forcing the client to disconnect first breaks the process in two steps which complicates it and makes it prone to race condition, i.e. client A (settings) disconnects the context, client B (connman) re-connects it back, client A fails to reset settings.

What about deactivating the context if anything significant (apn/username/password) has changed?